File: ai-demo-3/backend/app/services/session_manager.py
"""Session manager for handling multiple concurrent Nova Sonic sessions."""
import asyncio
import logging
from datetime import datetime
from typing import Dict, Optional
from app.services.nova_sonic_client import NovaSonicClient
from app.models.session import SessionStatus, SessionInfo
from app.config import settings

logger = logging.getLogger(__name__)


class SessionManager:
    """Manages multiple Nova Sonic client sessions."""
    
    def __init__(self):
        self.sessions: Dict[str, NovaSonicClient] = {}
        self.session_info: Dict[str, SessionInfo] = {}
        self._cleanup_task: Optional[asyncio.Task] = None
    
    async def create_session(
        self,
        system_prompt: Optional[str] = None,
        **kwargs
    ) -> tuple[str, NovaSonicClient]:
        """Create a new Nova Sonic session."""
        # Check concurrent session limit
        active_sessions = sum(
            1 for info in self.session_info.values()
            if info.status == SessionStatus.ACTIVE
        )
        
        if active_sessions >= settings.max_concurrent_sessions:
            raise ValueError(f"Maximum concurrent sessions ({settings.max_concurrent_sessions}) reached")
        
        # Create new client
        client = NovaSonicClient(system_prompt=system_prompt, **kwargs)
        
        try:
            await client.initialize_stream()
            
            # Store session
            session_id = client.session_id
            self.sessions[session_id] = client
            self.session_info[session_id] = SessionInfo(
                session_id=session_id,
                status=SessionStatus.ACTIVE,
                created_at=datetime.utcnow(),
                last_activity=datetime.utcnow()
            )
            
            logger.info(f"Session created: {session_id}")
            return session_id, client
            
        except Exception as e:
            logger.error(f"Failed to create session: {e}")
            raise
    
    def get_session(self, session_id: str) -> Optional[NovaSonicClient]:
        """Get an existing session."""
        return self.sessions.get(session_id)
    
    def get_session_info(self, session_id: str) -> Optional[SessionInfo]:
        """Get session information."""
        return self.session_info.get(session_id)
    
    async def end_session(self, session_id: str):
        """End a session and clean up resources."""
        client = self.sessions.get(session_id)
        if not client:
            logger.warning(f"Session not found: {session_id}")
            return
        
        try:
            await client.close()
            
            # Update session info
            if session_id in self.session_info:
                self.session_info[session_id].status = SessionStatus.ENDED
                self.session_info[session_id].last_activity = datetime.utcnow()
            
            # Remove from active sessions
            del self.sessions[session_id]
            
            logger.info(f"Session ended: {session_id}")
            
        except Exception as e:
            logger.error(f"Error ending session {session_id}: {e}")
            if session_id in self.session_info:
                self.session_info[session_id].status = SessionStatus.ERROR
    
    async def update_session_activity(self, session_id: str):
        """Update last activity timestamp for a session."""
        if session_id in self.session_info:
            self.session_info[session_id].last_activity = datetime.utcnow()
    
    async def cleanup_inactive_sessions(self):
        """Clean up sessions that have been inactive for too long."""
        now = datetime.utcnow()
        timeout_seconds = settings.session_timeout
        
        sessions_to_end = []
        for session_id, info in self.session_info.items():
            if info.status == SessionStatus.ACTIVE and info.last_activity:
                inactive_duration = (now - info.last_activity).total_seconds()
                if inactive_duration > timeout_seconds:
                    sessions_to_end.append(session_id)
        
        for session_id in sessions_to_end:
            logger.info(f"Cleaning up inactive session: {session_id}")
            await self.end_session(session_id)
    
    async def start_cleanup_task(self):
        """Start background task for cleaning up inactive sessions."""
        async def cleanup_loop():
            while True:
                try:
                    await asyncio.sleep(60)  # Check every minute
                    await self.cleanup_inactive_sessions()
                except asyncio.CancelledError:
                    break
                except Exception as e:
                    logger.error(f"Error in cleanup task: {e}")
        
        self._cleanup_task = asyncio.create_task(cleanup_loop())
        logger.info("Cleanup task started")
    
    async def stop_cleanup_task(self):
        """Stop background cleanup task."""
        if self._cleanup_task:
            self._cleanup_task.cancel()
            try:
                await self._cleanup_task
            except asyncio.CancelledError:
                pass
            logger.info("Cleanup task stopped")
    
    async def shutdown(self):
        """Shutdown all sessions and cleanup."""
        logger.info("Shutting down session manager")
        
        # Stop cleanup task
        await self.stop_cleanup_task()
        
        # Close all active sessions
        session_ids = list(self.sessions.keys())
        for session_id in session_ids:
            await self.end_session(session_id)
        
        logger.info("Session manager shutdown complete")


# Global session manager instance
session_manager = SessionManager()

File: ai-demo-3-nodejs/src/services/sessionManager.ts
/**
 * Session Manager for handling multiple concurrent Nova Sonic sessions
 */
import { NovaSonicClient, NovaSonicClientOptions } from './novaSonicClient';
import { SessionInfo, SessionStatus } from '../models/session';
import config from '../config';
import logger from '../logger';

export class SessionManager {
  private sessions: Map<string, NovaSonicClient>;
  private sessionInfo: Map<string, SessionInfo>;
  private activeStreams: Map<string, boolean>;
  private cleanupInterval?: NodeJS.Timeout;

  constructor() {
    this.sessions = new Map();
    this.sessionInfo = new Map();
    this.activeStreams = new Map();
  }

  async createSession(options: NovaSonicClientOptions = {}): Promise<[string, NovaSonicClient]> {
    // Check concurrent session limit
    const activeCount = Array.from(this.sessionInfo.values()).filter(
      (info) => info.status === SessionStatus.ACTIVE
    ).length;

    if (activeCount >= config.session.maxConcurrent) {
      throw new Error(
        `Maximum concurrent sessions (${config.session.maxConcurrent}) reached`
      );
    }

    // Create new client
    const client = new NovaSonicClient(options);

    try {
      await client.initializeStream();

      // Store session
      const sessionId = client.sessionId;
      this.sessions.set(sessionId, client);
      this.sessionInfo.set(sessionId, {
        sessionId,
        status: SessionStatus.ACTIVE,
        createdAt: new Date(),
        lastActivity: new Date(),
        audioBytesSent: 0,
        audioBytesReceived: 0,
        messageCount: 0,
      });

      logger.info(`Session created: ${sessionId}`);
      return [sessionId, client];
    } catch (error) {
      logger.error(`Failed to create session: ${error}`);
      throw error;
    }
  }

  getSession(sessionId: string): NovaSonicClient | undefined {
    return this.sessions.get(sessionId);
  }

  getSessionInfo(sessionId: string): SessionInfo | undefined {
    return this.sessionInfo.get(sessionId);
  }

  hasActiveStream(sessionId: string): boolean {
    return this.activeStreams.get(sessionId) || false;
  }

  markStreamActive(sessionId: string): void {
    this.activeStreams.set(sessionId, true);
    logger.debug(`Stream marked active for session: ${sessionId}`);
  }

  markStreamInactive(sessionId: string): void {
    this.activeStreams.delete(sessionId);
    logger.debug(`Stream marked inactive for session: ${sessionId}`);
  }

  async endSession(sessionId: string): Promise<void> {
    const client = this.sessions.get(sessionId);
    if (!client) {
      logger.warn(`Session not found: ${sessionId}`);
      return;
    }

    try {
      await client.close();

      // Update session info
      const info = this.sessionInfo.get(sessionId);
      if (info) {
        info.status = SessionStatus.ENDED;
        info.lastActivity = new Date();
      }

      // Clean up stream tracking
      this.markStreamInactive(sessionId);

      // Remove from active sessions
      this.sessions.delete(sessionId);

      logger.info(`Session ended: ${sessionId}`);
    } catch (error) {
      logger.error(`Error ending session ${sessionId}: ${error}`);
      const info = this.sessionInfo.get(sessionId);
      if (info) {
        info.status = SessionStatus.ERROR;
      }
    }
  }

  async updateSessionActivity(sessionId: string): Promise<void> {
    const info = this.sessionInfo.get(sessionId);
    if (info) {
      info.lastActivity = new Date();
    }
  }

  async cleanupInactiveSessions(): Promise<void> {
    const now = new Date();
    const timeoutMs = config.session.timeout * 1000;

    const sessionsToEnd: string[] = [];
    for (const [sessionId, info] of this.sessionInfo.entries()) {
      if (info.status === SessionStatus.ACTIVE && info.lastActivity) {
        const inactiveDuration = now.getTime() - info.lastActivity.getTime();
        if (inactiveDuration > timeoutMs) {
          sessionsToEnd.push(sessionId);
        }
      }
    }

    for (const sessionId of sessionsToEnd) {
      logger.info(`Cleaning up inactive session: ${sessionId}`);
      await this.endSession(sessionId);
    }
  }

  startCleanupTask(): void {
    this.cleanupInterval = setInterval(async () => {
      try {
        await this.cleanupInactiveSessions();
      } catch (error) {
        logger.error(`Error in cleanup task: ${error}`);
      }
    }, 60000); // Check every minute

    logger.info('Cleanup task started');
  }

  stopCleanupTask(): void {
    if (this.cleanupInterval) {
      clearInterval(this.cleanupInterval);
      logger.info('Cleanup task stopped');
    }
  }

  async shutdown(): Promise<void> {
    logger.info('Shutting down session manager');

    // Stop cleanup task
    this.stopCleanupTask();

    // Close all active sessions
    const sessionIds = Array.from(this.sessions.keys());
    for (const sessionId of sessionIds) {
      await this.endSession(sessionId);
    }

    logger.info('Session manager shutdown complete');
  }
}

// Global session manager instance
export const sessionManager = new SessionManager();

